src/NServiceBus.Transport.SQS/InputQueuePump.cs

```diff
@@ -308,69 +317,103 @@ internal async Task ProcessMessage(Message receivedMessage, CancellationToken me
             if (!IsMessageExpired(receivedMessage, transportMessage.Headers, messageId, CorrectClockSkew.GetClockCorrectionForEndpoint(awsEndpointUrl)))
             {
                 // here we also want to use the native message id because the core demands it like that
-                await ProcessMessageWithInMemoryRetries(transportMessage.Headers, nativeMessageId, messageBody, receivedMessage, messageProcessingCancellationToken).ConfigureAwait(false);
+                await InnerProcessMessage(transportMessage.Headers, nativeMessageId, messageBody, transportMessage.S3BodyKey, receivedMessage, messageProcessingCancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                await DeleteMessage(receivedMessage, transportMessage.S3BodyKey, messageProcessingCancellationToken).ConfigureAwait(false);
             }
-
-            // Always delete the message from the queue.
-            // If processing failed, the onError handler will have moved the message
-            // to a retry queue.
-            await DeleteMessage(receivedMessage, transportMessage.S3BodyKey, messageProcessingCancellationToken).ConfigureAwait(false);
         }
 
-        async Task ProcessMessageWithInMemoryRetries(Dictionary<string, string> headers, string nativeMessageId, byte[] body, Message nativeMessage, CancellationToken messageProcessingCancellationToken)
+        async Task InnerProcessMessage(Dictionary<string, string> headers, string nativeMessageId, byte[] body, string s3BodyKey, Message nativeMessage, CancellationToken messageProcessingCancellationToken)
```

Comment:
I'm thinking we could change this method to return a `bool` that this true if the message was processed successfully instead of deleting the message in this method. That would let us remove having to pass in the `s3BodyKey.